geomstats/learning/geodesic_regression.py

```diff
@@ -0,0 +1,381 @@
+
+import geomstats.backend as gs
+import geomstats.errors
+from geomstats.learning.frechet_mean import FrechetMean
+
+
+class GeodesicRegression(BaseEstimator):
+    r"""Geodesic Regression.
+
+    The generative model of the data is:
+        :math:`Z = Exp_{\beta_0}(\beta_1.X)` and :math:`Y = Exp_Z(\epsilon)`
+        where:
+        - :math:`Exp` denotes the Riemannian exponential,
+        - :math:`\beta_0` is called the intercept,
+        - :math:`\beta_1` is called the coefficient,
+        - :math:`\epsilon \sim N(0, 1)` is a standard Gaussian noise,
+        - :math:`X` is the input, :math:`Y` is the target.
+
+    The geodesic regression method:
+    - estimates :math:`\beta_0, \beta_1`,
+    - predicts :math:`\hat{y}` from input :math:`X`.
+
+    Parameters
+    ----------
+    space : Manifold
+        Manifold.
+    metric : RiemannianMetric
+        Riemannian metric.
+    center_X : bool
+        Subtract mean to X as a preprocessing.
+    method : str, {\'extrinsic\', \'riemannian\'}
+        Gradient descent method.
+        Optional, default: extrinsic.
+    max_iter : int
+        Maximum number of iterations for gradient descent.
+        Optional, default: 100.
+    learning_rate : float
+        Initial learning rate for gradient descent.
+        Optional, default: 0.1
+    tol : float
+        Tolerance for loss minimization.
+        Optional, default: 1e-5
+    verbose : bool
+        Verbose option.
+        Optional, default: False.
+    """
+
+    def __init__(
+            self, space, metric=None, center_X=True, method='extrinsic',
+            max_iter=100, learning_rate=.1, tol=1e-5, verbose=False):
+        if metric is None:
+            metric = space.metric
+        self.metric = metric
+        self.space = space
+        self.intercept_ = None
+        self.coef_ = None
+        self.center_X = center_X
+        self.mean_ = None
+        self.training_score_ = None
+        geomstats.errors.check_parameter_accepted_values(
+            method, 'method', ['extrinsic', 'riemannian'])
+        self.method = method
+        self.max_iter = max_iter
+        self.verbose = verbose
+        self.learning_rate = learning_rate
+        self.tol = tol
+
+    def _model(self, X, coef, intercept):
+        """Compute the generative model of the geodesic regression.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=[...,}]
+            Training input samples.
+        coef : array-like, shape=[..., dim]
+            Coefficient of the geodesic regression.
+        intercept : array-like, shape=[..., dim]
+            Intercept of the geodesic regression.
+
+        Returns
+        -------
+        _ : array-like, shape=[..., dim]
+            Value on the manifold output by the generative model.
+        """
+        X = X[:, None] if self.metric.default_point_type == 'vector' else\
+            X[:, None, None]
+        return self.metric.exp(X * coef[None], intercept)
```

Comment:
Does this method modify X? This might be confusing

Reply:
Good question, I'm not sure it does but just in case I'll rename X.